### PR TITLE
Multiscale Competitive CNN by mkosturek

### PR DESCRIPTION
**Pull request by mkosturek**
https://arxiv.org/abs/1511.05635

Implementation in Pytorch
Tested on MNIST & CIFAR 10

### CODE REVIEW

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    def forward(self, x):
+        outs = [bnorm(conv(x)) for conv, bnorm in zip(self.convs, self.bnorms)]
+        return self.maxout(outs)
+
+
+class CompetitiveMultiscaleCNN(nn.Module):
+
+    def __init__(self, in_channels):
+
```
Comment (secezar) [inline src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: I suggest removing empty lines between method definition and its implementation. :)

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        self.bnorms = [nn.BatchNorm2d(nb_kernels) for _ in kernels_sizes]
+        self.maxout = MaxOut()
+
+    def forward(self, x):
+        outs = [bnorm(conv(x)) for conv, bnorm in zip(self.convs, self.bnorms)]
+        return self.maxout(outs)
+
+
+class CompetitiveMultiscaleCNN(nn.Module):
+
```
Comment (secezar) [inline src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: I think we could add short docstring with paper arxiv link.

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            cur_max = torch.max(cur_max, inp)
+        return cur_max
+
+
+class ConvCompetitiveLayer(nn.Module):
+
+    def __init__(self, in_channels, nb_kernels, kernels_sizes):
+        super(ConvCompetitiveLayer, self).__init__()
+
+        self.convs = [nn.Conv2d(in_channels, nb_kernels, k_sz, padding=k_sz//2)
```
Comment (secezar) [inline src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: What does k_sz do? xd

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        self.block1_layer_2 = ConvCompetitiveLayer(192, 192, [1, 3])
+
+        self.block1_maxpool = nn.MaxPool2d(3)
+        self.block1_dropout = nn.Dropout2d()
+
+        self.block2_layer_0 = ConvCompetitiveLayer(192, 192, [1, 3, 5, 7])
+        self.block2_layer_1 = ConvCompetitiveLayer(192, 192, [1, 3])
+        self.block2_layer_2 = ConvCompetitiveLayer(192, 10, [1, 3])
+
+        self.block2_avgpool = nn.AvgPool2d(3)
```
Comment (kacper1095) [inline src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Not adapted to different input sizes

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+class ConvCompetitiveLayer(nn.Module):
+
+    def __init__(self, in_channels, nb_kernels, kernels_sizes):
+        super(ConvCompetitiveLayer, self).__init__()
+
+        self.convs = [nn.Conv2d(in_channels, nb_kernels, k_sz, padding=k_sz//2)
+                      for k_sz in kernels_sizes]
+
+        self.bnorms = [nn.BatchNorm2d(nb_kernels) for _ in kernels_sizes]
```
Comment (secezar) [inline src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: As you use only kernel_sizes length, maybe it could be better to write it like:
`[nn.BatchNorm2d(nb_kernels)]*len(kernel_sizes)`?

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+
+class MaxOut(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, inputs):
+        cur_max = inputs[0]
+        for inp in inputs[1:]:
```
Comment (secezar) [inline src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: I think that `input` and `current_max` names are better (I personally don't like abbreviations, because they can lead to code misunderstanding).

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            cur_max = torch.max(cur_max, inp)
```
Comment (secezar) [inline src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Are you sure that this function does not take Sequential object (like inputs) as an argument (you search max elementwise)? I'm curious and it is hard to read it from documentation.

**Review by secezar — CHANGES_REQUESTED** [review]

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 3 above)
Comment (PiotrGrzybowski) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: +1, maybe just kernel_size. I do not think the name is too long.

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 7 above)
Comment (mkosturek) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Nope, it takes two Tensors and compares them elementwise:

Documentation says:

> `torch.max(input, other, out=None) → Tensor`
Each element of the tensor input is compared with the corresponding element of the tensor other and an element-wise max is taken.
The shapes of input and other don’t need to match, but they must be broadcastable.

**Review by PiotrGrzybowski — CHANGES_REQUESTED** [review]

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 6 above)
Comment (mkosturek) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Name `input` overrides built-in python function - which I tried to avoid.
I wonder if the name `cur_max` isn't just right for such small scope (only 4-liner function).

Comment (secezar) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: As `input` name overriding is indeed bad, `current_max` makes it just no-brainer name and in general makes code reading faster.

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 5 above)
Comment (mkosturek) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Wouldn't it duplicate the same BatchNorm2d object multiple times instead of creating separate ones that should be optimised separately?

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 2 above)
Comment (mkosturek) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Will do!

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 1 above)
Comment (mkosturek) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Will do!

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 3 above)
Comment (mkosturek) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: as it says in line `23` - `for k_sz in kernel_sizes`...

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 5 above)
Comment (secezar) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Yes it does creat copies.

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py (hunk 3 above)
Comment (secezar) [reply src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py]: Maybe `k_s`? I just wanted to point that `z` comes from the middle of word.

**Review by secezar — APPROVED** [review]

**Review by PiotrGrzybowski — APPROVED** [review]

File: src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        self.block1_layer_2 = ConvCompetitiveLayer(192, 192, [1, 3])
+
+        self.block1_maxpool = nn.MaxPool2d(3)
+        self.block1_dropout = nn.Dropout2d()
+
+        self.block2_layer_0 = ConvCompetitiveLayer(192, 192, [1, 3, 5, 7])
+        self.block2_layer_1 = ConvCompetitiveLayer(192, 192, [1, 3])
+        self.block2_layer_2 = ConvCompetitiveLayer(192, 10, [1, 3])
+
+        self.block2_avgpool = nn.AvgPool2d(avg_pool_size)
```
Comment (kacper1095) [inline src/models/architectures/cnns/architectures/competitive_multiscale_cnn.py:65]: Still non adaptive pooling size. Now it requires from user to know the implementation of the model so he can put appropriate value in the constructor.
